components/webext-storage/src/error.rs

```diff
@@ -2,17 +2,46 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+use error_support::{ErrorHandling, GetErrorHandling};
 use interrupt_support::Interrupted;
+use std::fmt;
 
-#[derive(Debug)]
+/// Result enum for the public interface
+pub type ApiResult<T> = std::result::Result<T, WebExtStorageApiError>;
+/// Result enum for internal functions
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, Clone, Copy)]
 pub enum QuotaReason {
     TotalBytes,
     ItemBytes,
     MaxItems,
 }
 
+impl fmt::Display for QuotaReason {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            QuotaReason::ItemBytes => write!(f, "ItemBytes"),
+            QuotaReason::MaxItems => write!(f, "MaxItems"),
+            QuotaReason::TotalBytes => write!(f, "TotalBytes"),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum WebExtStorageApiError {
+    #[error("Unexpected webext-storage error: {reason}")]
+    UnexpectedError { reason: String },
+
+    #[error("Error parsing JSON data: {reason}")]
+    JsonError { reason: String },
+
+    #[error("Quota exceeded: {reason}")]
+    QuotaError { reason: String },
```

Comment:
This can probably stick with the enum rather than a string. It will mean declaring the enum in the UDL, but that seems fine.